components/script/dom/cookiestore.rs

```diff
@@ -0,0 +1,568 @@
+
+        // 3. If origin is an opaque origin, then return a promise rejected with a "SecurityError" DOMException.
+        if !origin.is_tuple() {
+            p.reject_error(Error::Security, can_gc);
+            return p;
+        }
+
+        // 4. Let url be settings’s creation URL.
+        // 5. Let domain be null.
+        // 6. Let path be "/".
+        // 7. Let sameSite be strict.
+        // 8. Let partitioned be false.
+        let cookie = Cookie::build((Cow::Owned(name.to_string()), Cow::Owned(value.to_string())))
+            .path("/")
+            .secure(true)
+            .same_site(SameSite::Strict)
+            .partitioned(false);
+        // TODO: This currently doesn't implement all the "set a cookie" steps which involves
+        // additional processing of the name and value
+
+        self.in_flight.borrow_mut().push_back(p.clone());
+
+        // 10. Run the following steps in parallel:
+        let _ = self
+            .global()
+            .resource_threads()
+            .send(CoreResourceMsg::SetCookieForUrlAsync(
+                self.store_id,
+                self.global().creation_url().clone(),
+                Serde(cookie.build()),
+                NonHTTP,
+            ));
+
+        // 11. Return p.
+        p
+    }
+
+    /// <https://cookiestore.spec.whatwg.org/#dom-cookiestore-set-options>
+    fn Set_(&self, options: &CookieInit, can_gc: CanGc) -> Rc<Promise> {
+        // 1. Let settings be this’s relevant settings object.
+        let global = self.global();
+
+        // 2. Let origin be settings’s origin.
+        let origin = global.origin();
+
+        // 5. Let p be a new promise.
+        let p = Promise::new(&global, can_gc);
+
+        // 3. If origin is an opaque origin, then return a promise rejected with a "SecurityError" DOMException.
+        if !origin.is_tuple() {
+            p.reject_error(Error::Security, can_gc);
+            return p;
+        }
+
+        // 4. Let url be settings’s creation URL.
+        let creation_url = global.creation_url();
+
+        // 6.1. Let r be the result of running set a cookie with url, options["name"], options["value"],
+        // options["expires"], options["domain"], options["path"], options["sameSite"], and options["partitioned"].
+        let cookie = Cookie::build((
+            Cow::Owned(options.name.to_string()),
+            Cow::Owned(options.value.to_string()),
+        ));
+        // TODO: This currently doesn't implement all the "set a cookie" steps which involves
+        // additional processing of the name and value
+
+        self.in_flight.borrow_mut().push_back(p.clone());
+
+        // 6. Run the following steps in parallel:
+        let _ = self
+            .global()
+            .resource_threads()
+            .send(CoreResourceMsg::SetCookieForUrlAsync(
+                self.store_id,
+                creation_url.clone(),
+                Serde(cookie.build()),
+                NonHTTP,
+            ));
```

Comment:
If it fails you probably want to. remove the inflight promise as well, and reject it with an error.

Reply:
Ah, yeah probably just going to make a small helper for this then. I wish there was a nice way to make this a type/lifetime error somehow.